debezium-core/src/main/java/io/debezium/pipeline/source/snapshot/incremental/AbstractIncrementalSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,388 @@
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void closeWindow(String id, OffsetContext offsetContext) throws InterruptedException {
+        context = (IncrementalSnapshotContext<T>) offsetContext.getIncrementalSnapshotContext();
+        if (!context.closeWindow(id)) {
+            return;
+        }
+        sendWindowEvents(offsetContext);
+    }
+
+    protected void sendWindowEvents(OffsetContext offsetContext) throws InterruptedException {
+        LOGGER.debug("Sending {} events from window buffer", window.size());
+        offsetContext.incrementalSnapshotEvents();
+        for (Object[] row : window.values()) {
+            sendEvent(dispatcher, offsetContext, row);
+        }
+        offsetContext.postSnapshotCompletion();
+        window.clear();
+        readChunk();
+    }
+
+    protected void sendEvent(EventDispatcher<T> dispatcher, OffsetContext offsetContext, Object[] row) throws InterruptedException {
+        context.sendEvent(keyFromRow(row));
+        offsetContext.event(context.currentDataCollectionId(), clock.currentTimeAsInstant());
+        dispatcher.dispatchSnapshotEvent(context.currentDataCollectionId(),
+                getChangeRecordEmitter(context.currentDataCollectionId(), offsetContext, row),
+                dispatcher.getIncrementalSnapshotChangeEventReceiver(dataListener));
+    }
+
+    /**
+     * Returns a {@link ChangeRecordEmitter} producing the change records for
+     * the given table row.
+     */
+    protected ChangeRecordEmitter getChangeRecordEmitter(T dataCollectionId, OffsetContext offsetContext,
+                                                         Object[] row) {
+        return new SnapshotChangeRecordEmitter(offsetContext, row, clock);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void processMessage(DataCollectionId dataCollectionId, Object key, OffsetContext offsetContext) throws InterruptedException {
+        context = (IncrementalSnapshotContext<T>) offsetContext.getIncrementalSnapshotContext();
+        if (context == null) {
+            return;
+        }
+        context.updateWindowState(offsetContext);
+        LOGGER.trace("Checking window for table '{}', key '{}', window contains '{}'", dataCollectionId, key, window);
+        if (!window.isEmpty()) {
+            if (context.deduplicationNeeded()) {
+                deduplicateWindow(dataCollectionId, key);
+            }
+            else {
+                sendWindowEvents(offsetContext);
+            }
+        }
+    }
+
+    private void deduplicateWindow(DataCollectionId dataCollectionId, Object key) {
+        if (!context.currentDataCollectionId().equals(dataCollectionId)) {
+            return;
+        }
+        if (key instanceof Struct) {
+            if (window.remove((Struct) key) != null) {
+                LOGGER.info("Removed '{}' from window", key);
+            }
+        }
+    }
+
+    protected abstract void emitWindowOpen() throws SQLException;
```

Comment:
A JavaDoc comment of this method (and the next one) would be nice, given their framework-kind nature.